### ISSUE TITLE
Using ${callsite} returns invalid classname "NLog.Extensions.Logging.NLogLogger.Log"

### ISSUE THREAD

**Issue by agentschmitt** [NLog/NLog.Web#206]
**Type**: Bug

- What is the current result?
2017-10-16 11:27:52.680 [5] INFO NLog.Extensions.Logging.NLogLogger.Log - application started

- What is the expected result?
2017-10-16 11:27:52.680 [5] INFO My.Web.Api.Startup- application started

- Are there any work arrounds? yes:

```c#
LogManager.AddHiddenAssembly(Assembly.Load(new AssemblyName("Microsoft.Extensions.Logging")));
LogManager.AddHiddenAssembly(Assembly.Load(new AssemblyName("Microsoft.Extensions.Logging.Abstractions")));
LogManager.AddHiddenAssembly(Assembly.Load(new AssemblyName("NLog.Extensions.Logging")));
```

I Have added this code to my application and then logging of classname works again, see code example below.

**Related Issue**:
https://github.com/NLog/NLog.Web/issues/200
https://github.com/NLog/NLog.Web/pull/202/files

It seems Nlog.Web Assembly is added to hidden Assemblys now, but this doesn't help.
It is required to add the NLog.Extensions.Logging + the Microsoft.Extensions.
I am also not sure it this bug is also in .NET Core 2 Console Applications, will have to test this.
If its there also, the fix wold be required in NLog.Extensions.Logging instead of Nlog.Web?

**NLog version**: (4.5.0 Beta7)
**NLog.Extensions.Logging version**: (1.0.0 RTM Beta7)
**NLog.Web.AspNetCore version**: (4.5.0) (Beta 2)

**Platform**:  .NET Core 2

**Current NLog config** (xml or C#, if relevant)

```xml
<?xml version="1.0" encoding="utf-8" ?>
<nlog xmlns="http://www.nlog-project.org/schemas/NLog.xsd"
      xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
      autoReload="true"
      internalLogLevel="Warn"
      internalLogFile="..\..\logs\internal-NLog-web.txt">

  <variable name="formattedMessage" value="${message}${onexception:${newline}${exception:format=ToString}}" />
  <variable name="defaultLayout" value="${date:format=yyyy-MM-dd HH\:mm\:ss.fff} [${threadid}] ${level:uppercase=true} ${callsite} - ${formattedMessage}" />
  <variable name="logFolder" value="..\..\..\..\..\logs" />

  <targets async="true">
    <target xsi:type="File" name="rootFileTarget" fileName="${logFolder}\web-${shortdate}.log" layout="${defaultLayout}" keepFileOpen="true" />
    <target xsi:type="File" name="msFileTarget" fileName="${logFolder}\ms-${shortdate}.log" layout="${defaultLayout}" keepFileOpen="true" />
  </targets>

  <rules>
    <logger name="Microsoft.*" minlevel="Debug" writeTo="msFileTarget" />
    <logger name="Microsoft.*" final="true" />
    <logger name="*" minlevel="Debug" writeTo="rootFileTarget" />
  </rules>
</nlog>
```
```c#
namespace My.Web.Api
{
    public static class Program
    {
        public static void Main(string[] args)
        {
            var logger = NLogBuilder.ConfigureNLog("NLog.config").GetCurrentClassLogger();

            try
            {
                BuildWebHost(args).Run();
            }
            catch (Exception e)
            {
                logger.Error(e, "application crashed");
                throw;
            }
        }

        public static IWebHost BuildWebHost(string[] args)
        {
            // fixes nlog callsite
            LogManager.AddHiddenAssembly(Assembly.Load(new AssemblyName("Microsoft.Extensions.Logging")));
            LogManager.AddHiddenAssembly(Assembly.Load(new AssemblyName("Microsoft.Extensions.Logging.Abstractions")));
            LogManager.AddHiddenAssembly(Assembly.Load(new AssemblyName("NLog.Extensions.Logging")));

            var hostBuilder = WebHost.CreateDefaultBuilder(args)
                .UseStartup<Startup>()
                .UseNLog();

            return hostBuilder.Build();
        }
    }

    public class Startup
    {
        private ILogger<Startup> Logger { get; }

        public Startup(ILogger<Startup> logger)
        {
            Logger = logger;
            Logger.LogInformation($"application started");
        }    
    }
}
```

**Comment by 304NotModified** [comment]
Thanks for reporting, will look at this.

**Comment by agentschmitt** [comment]
In some cases Assembly.Load for DotNet Core Framework Libs can make problems: https://github.com/dotnet/coreclr/issues/12668

Running our published application on an docker container with microsoft/aspnetcore (version 2.0.0),
the application is crashed by nlog.

There is a workaround though, that can be added before loading nlog:
`
            Thread.CurrentThread.CurrentUICulture = new CultureInfo("en-US");
            Thread.CurrentThread.CurrentCulture = new CultureInfo("en-US");
`

`Assert Failure
Expression: [Recursive resource lookup bug]
Description: Infinite recursion during resource lookup within System.Private.CoreLib. This may be a bug in System.Private.CoreLib, or potentially in certain extensibility points such as assembly resolve events or CultureInfo names. Resour
ce name: ArgumentNull_Generic
Stack Trace:
at System.SR.InternalGetResourceString(String key)
at System.SR.GetResourceString(String resourceKey, String defaultString)
at System.ArgumentNullException..ctor(String paramName)
at System.Runtime.Loader.AssemblyLoadContext.GetLoadContext(Assembly assembly)
at System.Reflection.Assembly.LoadFromResolveHandler(Object sender, ResolveEventArgs args)
at System.AppDomain.OnAssemblyResolveEvent(RuntimeAssembly assembly, String assemblyFullName)
at System.Reflection.RuntimeAssembly._nLoad(AssemblyName fileName, String codeBase, Evidence assemblySecurity, RuntimeAssembly locationHint, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntro
spection, Boolean suppressSecurityChecks, IntPtr ptrLoadContextBinder)
at System.Reflection.RuntimeAssembly.InternalGetSatelliteAssembly(String name, CultureInfo culture, Version version, Boolean throwOnFileNotFound, StackCrawlMark& stackMark)
at System.Resources.ManifestBasedResourceGroveler.GetSatelliteAssembly(CultureInfo lookForCulture, StackCrawlMark& stackMark)
at System.Resources.ManifestBasedResourceGroveler.GrovelForResourceSet(CultureInfo culture, Dictionary`2 localResourceSets, Boolean tryParents, Boolean createIfNotExists, StackCrawlMark& stackMark)
at System.Resources.ResourceManager.InternalGetResourceSet(CultureInfo requestedCulture, Boolean createIfNotExists, Boolean tryParents, StackCrawlMark& stackMark)
at System.Resources.ResourceManager.InternalGetResourceSet(CultureInfo culture, Boolean createIfNotExists, Boolean tryParents)
at System.Resources.ResourceManager.GetString(String name, CultureInfo culture)
at System.SR.InternalGetResourceString(String key)
at System.SR.GetResourceString(String resourceKey, String defaultString)
at System.ArgumentNullException..ctor(String paramName)
at System.Runtime.Loader.AssemblyLoadContext.GetLoadContext(Assembly assembly)
at System.Reflection.Assembly.LoadFromResolveHandler(Object sender, ResolveEventArgs args)
at System.AppDomain.OnAssemblyResolveEvent(RuntimeAssembly assembly, String assemblyFullName)
at System.Reflection.RuntimeAssembly._nLoad(AssemblyName fileName, String codeBase, Evidence assemblySecurity, RuntimeAssembly locationHint, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntro
spection, Boolean suppressSecurityChecks, IntPtr ptrLoadContextBinder)
at System.Reflection.RuntimeAssembly.InternalGetSatelliteAssembly(String name, CultureInfo culture, Version version, Boolean throwOnFileNotFound, StackCrawlMark& stackMark)
at System.Resources.ManifestBasedResourceGroveler.GetSatelliteAssembly(CultureInfo lookForCulture, StackCrawlMark& stackMark)
at System.Resources.ManifestBasedResourceGroveler.GrovelForResourceSet(CultureInfo culture, Dictionary`2 localResourceSets, Boolean tryParents, Boolean createIfNotExists, StackCrawlMark& stackMark)
at System.Resources.ResourceManager.InternalGetResourceSet(CultureInfo requestedCulture, Boolean createIfNotExists, Boolean tryParents, StackCrawlMark& stackMark)
at System.Resources.ResourceManager.InternalGetResourceSet(CultureInfo culture, Boolean createIfNotExists, Boolean tryParents)
at System.Resources.ResourceManager.GetString(String name, CultureInfo culture)
at System.SR.InternalGetResourceString(String key)
at System.SR.GetResourceString(String resourceKey, String defaultString)
at System.Runtime.Loader.AssemblyLoadContext.ResolveUsingEvent(AssemblyName assemblyName)
at System.Runtime.Loader.AssemblyLoadContext.ResolveUsingResolvingEvent(IntPtr gchManagedAssemblyLoadContext, AssemblyName assemblyName)
at System.Reflection.RuntimeAssembly._nLoad(AssemblyName fileName, String codeBase, Evidence assemblySecurity, RuntimeAssembly locationHint, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntro
spection, Boolean suppressSecurityChecks, IntPtr ptrLoadContextBinder)
at System.Reflection.RuntimeAssembly.InternalLoadAssemblyName(AssemblyName assemblyRef, Evidence assemblySecurity, RuntimeAssembly reqAssembly, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIn
trospection, IntPtr ptrLoadContextBinder)
at System.Reflection.Assembly.Load(AssemblyName assemblyRef)
at NLog.Web.NLogBuilder.ConfigureHiddenAssemblies()
at NLog.Web.NLogBuilder.ConfigureNLog(LoggingConfiguration configuration)
at My.Web.Api.Program.Main(String[] args) in D:\WindowsBuildAgent3_Workspace\dcbf3a7ed9ca02a0\Sources\Web\Api\Program.cs:line 14`

**Comment by 304NotModified** [comment]
> eg. Loading Assembly by getting it from an type as done for "NLog.Extensions.Logging":
> LogManager.AddHiddenAssembly(typeof(ConfigureExtensions).GetTypeInfo().Assembly);

if this fixed the problem, then I think it's a good change. Could you do a PR for this?

**Comment by 304NotModified** [comment]
I see in https://github.com/dotnet/coreclr/issues/12668 , this is Fixed in 2.0?

**Comment by 304NotModified** [comment]
are you using .NET Core 2 RTM? Or a preview/beta?

**Comment by 304NotModified** [comment]
or this is probably dotnet/coreclr#14033

**Comment by 304NotModified** [comment]
closing this in favor of https://github.com/NLog/NLog.Web/issues/210

**Comment by agentschmitt** [comment]
I am using .NET Core 2.0.0 RTM.
It seems the bug was fixed already but not included in 2.0.0.
When 2.0.2 gets release it should be solved...
